### DISCUSSION TITLE
Need help with setup on existing Metro project

### DISCUSSION THREAD

**Discussion by Rizzaxc** [tamagui/tamagui#487] [Q&A] [answered]
Hi, I'm trying to switch my UI framework to Tamagui but it's been immensely difficult. 
Some parts of my app break completely even though there's no Typescript error.

![image](https://user-images.githubusercontent.com/33868883/210321645-e18cf153-793e-4d13-86d0-fc120743b7a6.png)

My custom component, which lives at `root/src/styles`, while can render, generates this warning. This component inherits from another library.

![image](https://user-images.githubusercontent.com/33868883/210321908-b30106b2-2a3c-4a84-ae5e-4e239b00b88f.png)

Some styling in the custom component doesn't make it to the simulator (sorry cant recall which).

My setup:

- `babel.config.js` at app root
```
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
  plugins: [
    [
      'module-resolver',
      {
        root: ['./src/'],
        extensions: ['.ios.js', '.android.js', '.js', '.json', '.ts', '.tsx'],
      },
    ],
    'react-native-reanimated/plugin',
    [
      '@tamagui/babel-plugin',
      {
        components: ['tamagui'],
        config: './src/styles/tamagui.config.ts',
        importsWhitelist: ['constants.js', 'colors.js'],
        logTimings: true,
        disableExtraction: process.env.NODE_ENV === 'development',
      },
    ],
    [
      'transform-inline-environment-variables',
      {
        include: 'TAMAGUI_TARGET',
      },
    ],
  ],
};

```

I'm aware I'm using `module-resolver` so have tried switching it around but the problems persist.

- `tamagui.config.ts` lives in the `root/src/styles` folder. My app entry already has Provider with the correct config path supplied
```
import {config} from '@tamagui/config-base';

import {createTamagui, createTokens} from 'tamagui';
import {shorthands} from '@tamagui/shorthands';

export const tokens = createTokens({
  size: config.tokens.size,
  space: config.tokens.space,
  zIndex: config.tokens.zIndex,
  radius: {
    ...config.tokens.radius,
    // custom stuff
  },
  color: {
    ...config.tokens.color,
    // custom stuff
  },
});

const tamaguiConfig = createTamagui({
  ...config,
  tokens,
  shorthands,
  themes: {
    light: {
      ...config.themes.light,
      // custom stuff
    },
  },
});

export type TamaguiConfig = typeof tamaguiConfig;
declare module 'tamagui' {
  interface TamaguiCustomConfig extends TamaguiConfig {}
}
export default tamaguiConfig;

``` 

I also have `TAMAGUI_TARGET=native` in my build commands

**Comment by iBotPeaches** [comment]
Not sure I can give you an easy answer, but we discovered in a pure RN application - a bit tougher than Expo. So for the fonts - they are normally assumed you have expo and thus `expo-font` to handle the loading of fonts.

If you are using Metro - I'm assuming you'll want to use `react-native-asset` and load the Inter font manually. You can grab font files here - https://github.com/tamagui/tamagui/tree/master/packages/font-inter . That package will handle with `react-native link` used to do and install the font respectively for iOS/Android. 

If you are using Expo - you can see this handled inline - https://github.com/tamagui/tamagui/blob/master/apps/kitchen-sink/src/App.tsx#L11

**Reply by Rizzaxc** [reply]
it's a shame. the promises (if true) are very tempting but I've met so many footguns/ undoc-ed behaviors that require me to dig into library code

**Reply by LunatiqueCoder** [reply]
Indeed, setting it up is not as straight forward, but after I managed to overcome all issues, it was worth it! Great library!

**Comment by LunatiqueCoder** [comment, marked as answer]
1. Create a `react-native.config.js` on the root folder.
2. Paste the following: 
```javascript
module.exports = {
  project: {
    ios: {},
    android: {},
  },
  assets: [
    './node_modules/@tamagui/font-inter/otf',
    './node_modules/@tamagui/font-silkscreen/files',
  ],
};
```
3. Run `npx react-native-asset`

Let me know if that works. Cheers.

**Reply by Rizzaxc** [reply]
it works for the font issues. thanks a bunch. but there's still the `ref` thing which I believe indicates I've not set up the compiler correctly

**Reply by iBotPeaches** [reply]
@Rizzaxc for the ref issue, could you double check your lock file that you only have one version of `react` and `react-dom` installed? Saw a similar error to that when we had two `react-dom` 

If you don't - you'll have to setup resolutions to force a version

**Reply by LunatiqueCoder** [reply]
I managed to configure it without a monorepo and without Expo. Just vanilla React Native and Next.js, everything in one app. For now, everything works great! This library seems amazing! 

Let me know if you want me to explain you step by step how I did it.

**Reply by Rizzaxc** [reply]
@criszz77 that'd be cool, though right now I've chosen to ignore the ref warnings since I'll rewrite some parts around the custom component. I suggest you take it a step further and make a PR upstream with your sample.

**Reply by Rizzaxc** [reply]
> @Rizzaxc for the ref issue, could you double check your lock file that you only have one version of `react` and `react-dom` installed? Saw a similar error to that when we had two `react-dom`
> 
> If you don't - you'll have to setup resolutions to force a version

unfortunately doesn't do anything. i've tried both with yarn-resolutions and npm-overrides

**Comment by LunatiqueCoder** [comment]
### TLDR - **Bare React Native + Next.js + Tamagui**   - no monorepo
While it's still a work in progress, you can try it out: 

<pre>npx react-native init <b>MyAppName</b> --template @criszz77/luna</pre>

And you can see the source code here: https://github.com/criszz77/luna

Implementation
---
So I basically started from [🌘 Luna](https://github.com/criszz77/luna) which is my React Native module that tries to implement a simple react-native-web starter. In order to [handle microbrowsers well](https://github.com/criszz77/luna/issues/46) I started [integrating Next.js](https://github.com/criszz77/luna/pull/60). From here, adding Tamagui was fairly simple. 

❗️ **Caveats**: I've patched `metro-react-native-babel-transformer` to accept a `babel.config.native.js` file that wouldn't interfere with Next.js. Not sure how problematic this would be, guess I'll find out in the future.

**EDIT:** I just realised that you're trying to integrate Tamagui in an existing app.... I'm sorry, I thought you're trying to start a new project 🙈 I hope this still helps you one way or another
